pkg/controller/sparkapplication/controller.go

```diff
@@ -258,91 +259,88 @@ func (c *Controller) processNextItem() bool {
-				sparkApplicationID: getSparkApplicationID(pod),
-			}
-			if pod.Status.Phase == apiv1.PodSucceeded || pod.Status.Phase == apiv1.PodFailed {
-				currentDriverState.completionTime = metav1.Now()
-			}
-			// Fetch container ExitCode/Reason if Pod Failed.
-			if pod.Status.Phase == apiv1.PodFailed {
-				if len(pod.Status.ContainerStatuses) > 0 && pod.Status.ContainerStatuses[0].State.Terminated != nil {
-					terminatedState := pod.Status.ContainerStatuses[0].State.Terminated
-					currentDriverState.err = fmt.Errorf("driver pod failed with ExitCode: %d, Reason: %s", terminatedState.ExitCode, terminatedState.Reason)
-				} else {
-					currentDriverState.err = fmt.Errorf("driver container status missing.")
-				}
-			}
+			driverPod = pod
 			break
 		}
 	}
+	if driverPod == nil {
+		// Driver Pod was not found in Informer cache. Try to fetch it from APIServer.
+		glog.Infof("driver not found for SparkApplication: %s/%s. Trying to fetch driver from APIServer.", app.Namespace, app.Name)
+		driverPod, err = c.getDriverPodFromApiServer(app)
+		if err != nil {
+			return err
+		}
+	}
 
-	if currentDriverState != nil {
-		newState := driverPodPhaseToApplicationState(currentDriverState.podPhase)
+	if driverPod != nil {
+		newState := driverPodPhaseToApplicationState(driverPod.Status.Phase)
 		// Only record a driver event if the application state (derived from the driver pod phase) has changed.
 		if newState != app.Status.AppState.State {
-			c.recordDriverEvent(app, currentDriverState.podPhase, currentDriverState.podName)
+			c.recordDriverEvent(app, driverPod.Status.Phase, driverPod.Name)
 		}
 		if newState != v1beta1.UnknownState {
-			app.Status.DriverInfo.PodName = currentDriverState.podName
-			app.Status.SparkApplicationID = currentDriverState.sparkApplicationID
-			if currentDriverState.nodeName != "" {
-				if nodeIP := c.getNodeIP(currentDriverState.nodeName); nodeIP != "" {
+			app.Status.DriverInfo.PodName = driverPod.Name
+			app.Status.SparkApplicationID = getSparkApplicationID(driverPod)
+			if driverPod.Spec.NodeName != "" {
+				if nodeIP := c.getNodeIP(driverPod.Spec.NodeName); nodeIP != "" {
 					app.Status.DriverInfo.WebUIAddress = fmt.Sprintf("%s:%d", nodeIP, app.Status.DriverInfo.WebUIPort)
 				}
 			}
-			if app.Status.TerminationTime.IsZero() && !currentDriverState.completionTime.IsZero() {
-				app.Status.TerminationTime = currentDriverState.completionTime
+			if app.Status.TerminationTime.IsZero() && driverPod.Status.Phase == apiv1.PodSucceeded || driverPod.Status.Phase == apiv1.PodFailed {
+				app.Status.TerminationTime = metav1.Now()
 			}
 		}
-		// Expose any errors to the users.
-		if currentDriverState.err != nil {
-			app.Status.AppState.ErrorMessage = currentDriverState.err.Error()
+		// Fetch container ExitCode/Reason if Pod Failed.
+		if driverPod.Status.Phase == apiv1.PodFailed {
+			if len(driverPod.Status.ContainerStatuses) > 0 && driverPod.Status.ContainerStatuses[0].State.Terminated != nil {
```

Comment:
Let's remove `driverPod.Status.ContainerStatuses[0].State.Terminated != nil` from this if check and do the check after line 331 as:
```
if terminatedState != nil {...}
```